packages/world/src/systems/BuildSystem.sol

```diff
@@ -151,56 +157,36 @@ contract BuildSystem is System {
     internal
   {
     for (uint256 i = 0; i < coords.length; i++) {
-      Vec3 coord = coords[i];
-
-      (ProgramId program, EntityId target) = _getHookTarget(coord);
-
       if (ctx.buildType == ObjectTypes.ForceField) {
-        require(!target._exists(), "Force field overlaps with another force field");
-        ForceFieldUtils.setupForceField(base, coord);
+        (EntityId existing,) = ForceFieldUtils.getForceField(coords[i]);
+        require(!existing._exists(), "Force field overlaps with another force field");
+        ForceFieldUtils.setupForceField(base, coords[i]);
         // Return early as it is not possible for a new forcefield to have a program
         return;
       }
 
+      (ProgramId program, EntityId target, EnergyData memory energyData) = ForceFieldUtils.getHookTarget(coords[i]);
+
       if (!program.exists()) {
         continue;
       }
 
-      program.hook({ caller: ctx.caller, target: target, revertOnFailure: true, extraData: extraData }).onBuild({
+      program.hook({ caller: ctx.caller, target: target, revertOnFailure: energyData.energy > 0, extraData: extraData })
+        .onBuild({
         entity: base,
-        coord: coord,
+        coord: coords[i],
         slotType: ctx.slotType,
         objectType: ctx.buildType,
         orientation: ctx.orientation
       });
     }
   }
-
-  function _getHookTarget(Vec3 coord) internal returns (ProgramId, EntityId) {
-    (EntityId forceField, EntityId fragment) = ForceFieldUtils.getForceField(coord);
-    if (!forceField._exists()) {
-      return (ProgramId.wrap(0), forceField);
-    }
-
-    EnergyData memory machineData = updateMachineEnergy(forceField);
-    if (machineData.energy == 0) {
-      return (ProgramId.wrap(0), forceField);
-    }
-
-    // We know fragment is active because its forcefield exists, so we can use its program
-    ProgramId program = fragment._getProgram();
-    if (program.exists()) {
-      return (program, fragment);
-    }
-
-    return (forceField._getProgram(), forceField);
-  }
 }
 
 library BuildLib {
   function _executeBuild(BuildContext memory ctx) public returns (EntityId, Vec3[] memory) {
-    (ctx.callerEnergy,) = transferEnergyToPool(ctx.caller, Math.min(ctx.callerEnergy, BUILD_ENERGY_COST));
```

Comment:
hmm didnt we need to set this on the build ctx?

Reply:
no, we only need it here, so trying to not mutate the ctx if possible